temporal-sdk/src/main/java/io/temporal/failure/DefaultFailureConverter.java

```diff
@@ -0,0 +1,383 @@
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class DefaultFailureConverter implements FailureConverter {
+
+  private static final Logger log = LoggerFactory.getLogger(DefaultFailureConverter.class);
+
+  public static final String JAVA_SDK = "JavaSDK";
+
+  private static final String ENCODED_FAILURE_MESSAGE = "Encoded failure";
+  private static final String STACK_TRACE_KEY = "stack_trace";
+  private static final String MESSAGE_KEY = "message";
+
+  private static final Type HASH_MAP_STRING_STRING_TYPE =
+      new TypeToken<HashMap<String, String>>() {}.getType();
+
+  /**
+   * Stop emitting stack trace after this line. Makes serialized stack traces more readable and
+   * compact as it omits most of framework level code.
+   */
+  private static final ImmutableSet<String> CUTOFF_METHOD_NAMES =
+      ImmutableSet.of(
+          "io.temporal.internal.worker.POJOActivityImplementationFactory$POJOActivityImplementation.execute",
+          "io.temporal.internal.sync.POJOWorkflowTaskHandler$POJOWorkflowImplementation.execute");
+
+  /** Used to parse a stack trace line. */
+  private static final String TRACE_ELEMENT_REGEXP =
+      "((?<className>.*)\\.(?<methodName>.*))\\(((?<fileName>.*?)(:(?<lineNumber>\\d+))?)\\)";
+
+  private static final Pattern TRACE_ELEMENT_PATTERN = Pattern.compile(TRACE_ELEMENT_REGEXP);
+
+  private boolean encodeDefaultAttributes;
+
+  public DefaultFailureConverter() {
+    this(false);
+  }
+
+  public DefaultFailureConverter(boolean encodeDefaultAttributes) {
+    this.encodeDefaultAttributes = encodeDefaultAttributes;
+  }
+
+  @Override
+  public RuntimeException failureToException(Failure failure, DataConverter dataConverter) {
+    if (failure == null) {
+      return null;
+    }
+    if (failure.hasEncodedAttributes()) {
+      Payload encodedAttributesPayload = failure.getEncodedAttributes();
+      Map<String, String> encodedAttributes =
+          DefaultDataConverter.STANDARD_INSTANCE.fromPayload(
+              encodedAttributesPayload, HashMap.class, HASH_MAP_STRING_STRING_TYPE);
+      failure =
+          failure.toBuilder()
+              .setStackTrace(encodedAttributes.get(STACK_TRACE_KEY))
+              .setMessage(encodedAttributes.get(MESSAGE_KEY))
+              .build();
+    }
+    RuntimeException result = failureToExceptionImpl(failure, dataConverter);
+    if (result instanceof TemporalFailure) {
+      ((TemporalFailure) result).setFailure(failure);
+    }
+    if (failure.getSource().equals(JAVA_SDK) && !failure.getStackTrace().isEmpty()) {
+      StackTraceElement[] stackTrace = parseStackTrace(failure.getStackTrace());
+      result.setStackTrace(stackTrace);
+    }
+    return result;
+  }
+
+  private RuntimeException failureToExceptionImpl(Failure failure, DataConverter dataConverter) {
```

Comment:
I'm hesitant here on either this method (and the equivalent one below) should be `protected`, so that users may more easily redefine only the "decision chain", while reusing the `encoded_attributes` and stack trace bits below.

On the opposite side, we could make this class `final` and encourage customization by decorator paradigm instead.

I personally have a slight preference for the second approach, for it promote a safer customization mechanism for the vast majority of users (ie. users are more likely to convert their custom exceptions to Temporal exceptions). I however expect diverging opinions on that question.

Reply:
> so that users may more easily redefine only the "decision chain", while reusing the encoded_attributes and stack trace bits below.

Nah, IMO these failure converters need to be wrapped not inherited IMO. I'd consider making this whole class "final" but that is a bit much :-). IMO users need to delegate to other failure converters and their entire behavior or not.

Reply:
Agree with Chad. It's definitely not a piece a user should ever customize. If users really want to implement their CORE logic of FailureConvert - they should be forced to implement their own from scratch and take full responsibility for it. Let's not provide an easy way to hijack the core-core logic as it's intended to be overridable. It really is not.

Reply:
Great! Then I'm actually making this class final. There is just no reason to extend this class rather than implements the interface.

Reply:
I think that's a bit much unless we change `DefaultDataConverter` to be final too. I think we can discourage inheritance without preventing it. While I generally can understand the final-by-default stance, that is almost a project-wide design decision and I think we generally aren't adding final to all of our classes (@Spikhalskiy correct me if I'm wrong)